### PR TITLE
10684 recover source id in case of failure

### LINKED ISSUE #10684: Try to not expose layers pointing to non-existant nodes
As we have talked, we will check if there is a broken layer pointing to a non-valid (ghost?) node. If so, we will "recover" the layer pointing to the previous valid source. If there is no previous source, because that layer only had one node, that layer will not be returned. Also, we want to show the error with a notification, so the user at least will notice why that node/layer is not present.

@gfiorav is making his magic, stay tuned.

cc @CartoDB/support @CartoDB/solutions 

Thanks @xavijam! The way I see it (just to expand a bit):

# Synopsis

We want to fix the inconsistencies between layer's `source_id` and the analysis definition. So it's 2 problems to be solved: 

1. Detect inconsistencies.
2. Fix inconsistencies.

## Detect inconsistencies 

The problem is that when you add/remove/edit a `source_id`, two objects need to be updated:

+ Layer
+ Analysis definition

In that event, if you want to "correct" any inconsistencies, you can do it:

1. When updating a layer
2. When updating a definition
3. When presenting a layer

The problem with 1 and 2 is that it's likely that you'll wan't to validate the integrity of a layer or definition when the other part hasn't yet been updated. We can solve this with any regular transaction id protocol-like fix, but it's too much since this is a temporal fix.

So we're left with 3. 3 is a pretty bad solution since it implies modifying things (altering or even destroying Layers based on a definition) on presentation. Since it's a temporal fix, this concern is mitigated. 

## Fix inconsistencies 

The easiest way to do it until the analysis logic lives in the back (and all of this is unnecessary) is to send the previous `source_id` in the layer options. That way when an inconsistency is detected, we can switch back to the previous `source_id`. This won't work for the "drag" case but it's better than nothing.

### PR DESCRIPTION
**Pull request by gfiorav**
Fixes #10684 

## Motivation

When front does any operation that changes the analyses definition it has to perform 2 operations:

+ Put to layer with new source
+ Put to definition with new definition

Either arrive in an indeterminate order and more often than we'd like, the order is incorrect or one request is lost. This is bound to be avoided when logic for nodes lives in back, but as for now, I present this hack to validate sources upon presentation.

I choose presentation because it's the last chance for a source to be valid. Up to that point, a definition update could still make the source valid. 

## Synopsis

Every time the `source` of a layer is changed, the previous value is kept. When presenting the layer, if the current `source` is deemed inconsistent (i.e.: it doesn't point to any analysis in the definition because that other update failed or any other reason), the previous `source` is restored.

If no previous `source` is found, a new one is inferred looking for the highest `source` number for the letter or if that fails simply decreasing the number by 1. 

If the resulting number is inviable (a-1), a source node with that letter is created if no other source node for that letter is found, and finally the layer is pointed to it.

So, in short, we try to recover from backup, infer looking at other nodes with the same letter or parse in that order of preference.

Please CR @juanignaciosl

### CODE REVIEW

File: spec/models/carto/layer_spec.rb [hunk 1]
```diff
@@ -82,5 +83,83 @@
...
+        @layer.options['previous_source'].should be_nil
+
+        @layer.has_valid_source?.should be_false
+        @layer.attempt_source_fix
+
+        @layer.source.should eq "#{new_letter}0"
+      end
+
+      it 'recognizes a valid source' do
+        @layer.has_valid_source?.should be_true
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: This test should first make explicit why the source is actually valid.

File: app/controllers/carto/api/vizjson3_presenter.rb [hunk 2]
```diff
@@ -325,6 +325,13 @@ def to_vizjson
         if @layer.base?
           vizjson.merge(as_base)
         else
+          # TODO: This code is a hack to fix ghost nodes before presenting
```
Comment (juanignaciosl) [inline app/controllers/carto/api/vizjson3_presenter.rb]: GitHub links are often useful to provide more context, but it's a matter of taste, don't take this into account if you want.

File: app/models/carto/layer.rb [hunk 3]
```diff
@@ -241,8 +241,73 @@ def category
...
+    def has_valid_source?
+      analysis_node.present?
+    end
+
+    def source
+      options['source']
+    end
+
+    def source=(source)
+      backup_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: As `backup_source` is a one-liner, private method, used only once, I'd inline it.

File: app/models/carto/layer.rb [hunk 4]
```diff
@@ -241,8 +241,73 @@ def category
...
+      options['source'] = source
+    end
+
+    def attempt_source_fix
+      self.source = previous_source || inferred_source || parsed_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Just in case you read this before my test comments, I'm not sure whether `previous_source` check should be first.

File: app/models/carto/layer.rb [hunk 5]
```diff
@@ -241,8 +241,73 @@ def category
...
+    end
+
     private
 
+    def inferred_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: What do you think about renaming this to `last_analysis_node_id`

File: app/models/carto/layer.rb [hunk 6]
```diff
@@ -241,8 +241,73 @@ def category
...
+      return unless source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: I don't know if this would make sense, but this method works if `source == nil` but `options['letter'] != nil`, so maybe this check is not complete.

File: app/models/carto/layer.rb [hunk 7]
```diff
@@ -241,8 +241,73 @@ def category
...
+
+        letter_analysis_nodes = analysis_nodes.select do |analysis_node|
+          analysis_node.id.first == letter
+        end
+
+        letter_analysis_nodes.map(&:id).sort.last
+      end
+    end
+
+    def parsed_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Even though you parse the source to get the id, that's the _how_, not the _what_. I _think_ that this returns the next-to-last id, isn't it? Maybe that would be a better name.

I've done another related comment in the test, sorry for the duplication.

File: spec/models/carto/layer_spec.rb [hunk 8]
```diff
@@ -82,5 +83,83 @@
         @layer.send(:affected_table_names, sql).should =~ ["cartodb.cdb_tablemetadata", "public.spatial_ref_sys"]
       end
     end
+
+    describe 'source_id recovery' do
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: PS3: I miss a better naming that would make tracing strategies (`previous_source || inferred_source || parsed_source`) with the related tests easier. Minor, though.

File: app/models/carto/layer.rb [hunk 9]
```diff
@@ -241,8 +241,73 @@ def category
...
+
+      "#{letter}#{number > 0 ? number - 1 : 0}"
+    end
+
+    def backup_source
+      self.previous_source = source if source.present?
+    end
+
+    def previous_source
+      options['previous_source']
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Just recovering the previous one, without any further check, is enough?

File: spec/models/carto/layer_spec.rb [hunk 10]
```diff
@@ -82,5 +83,83 @@
...
+        @layer.source.should eq bad_source
+        @layer.options['previous_source'] = 'super fake'
+
+        @layer.has_valid_source?.should be_false
+        @layer.attempt_source_fix
+
+        @layer.source.should eq 'super fake'
+      end
+
+      it 'looks for highest same letter analysis when no backup' do
```
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Very nice question indeed, I hoped it'd inspire some debate.

Since the backup is performed automatically and playing with the backup itself is discouraged via private methods, I think it's trustworthy. My other argument is that it's faster to fetch the backup than do all the definition parsing.

File: spec/models/carto/layer_spec.rb [hunk 11]
```diff
@@ -82,5 +83,83 @@
...
+        @layer.source.should eq bad_source
+        @layer.options['previous_source'].should be_nil
+
+        @layer.has_valid_source?.should be_false
+        @layer.attempt_source_fix
+
+        @layer.source.should eq @analysis.natural_id
+      end
+
+      it 'tries parsing when no analysis with same letter nor backup' do
```
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Yeah, I've stumbled a bit with naming I agree 😆

File: spec/models/carto/layer_spec.rb [hunk 12]
```diff
@@ -82,5 +83,83 @@
...
+        new_source = "#{new_letter}7"
+
+        @layer.options['source'] = new_source
+        @layer.source.should eq new_source
+        @layer.options['previous_source'].should be_nil
+
+        @layer.has_valid_source?.should be_false
+        @layer.attempt_source_fix
+
+        @layer.source.should eq "#{new_letter}6"
```
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Well its not a number really, it's the same as all the `'manolo'` test values. Magic manolo sounds nice though 😛 

Any number above 0 is good enough, I'll try to make it obvious via variable naming.

File: spec/models/carto/layer_spec.rb (hunk 1 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: It's valid because it's pointing to a valid node (see before all)

File: app/models/carto/layer.rb (hunk 5 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: Naming is tricky here. There doesn't seem to be a winner but for sure it's more descriptive than `inferred_source`. I'll think of something!

File: app/models/carto/layer.rb (hunk 6 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: Good point. It's weird that you'd try this without no `source` at all, but why not. I'll change it!

File: app/models/carto/layer.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: Well, you could get defensive here but there's no reason for previous_source to be corrupted. It's also true that since we don't JSON validate layer#options someone could potentially get stuff in there ( :trollface: )

File: spec/models/carto/layer_spec.rb (hunk 10 above)
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: Yep, I guess that we're sure that the backup id is trustworthy. Nevertheless, can we guarantee that the backup id is coherent (or, at least, more coherent than the other strategies) when the id is fixed? As this is a consistency issue between two sides (stored id and actual nodes) I'm not even sure if there's any actual scenario where backup is better (or even valid).

File: spec/models/carto/layer_spec.rb (hunk 12 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: hehe! I'll pick a random number created than 0 for the reader to understand it's random. Good point!

File: app/models/carto/layer.rb (hunk 9 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: xD

I wasn't wondering about the corruption of the stored data but its validity. We know that `previous_source` was valid but can we know that it's still valid when it's recovered?

File: spec/models/carto/layer_spec.rb (hunk 8 above)
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: ^ I don't get your comment here.

File: spec/models/carto/layer_spec.rb (hunk 10 above)
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: It sounds really good! Does it make sense for the scenarios that this is meant to fix?

File: spec/models/carto/layer_spec.rb [hunk 13]
```diff
@@ -82,5 +83,83 @@
...
+      end
+
+      it 'never parses to < 0' do
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: Ok, I think that I get it.

File: spec/models/carto/layer_spec.rb (hunk 1 above)
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: Uhm, well. Move it to the top so it's the first one at least, please :_)

File: app/models/carto/layer.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: I'm working on it. It's related to the the previous proposal of switching strategy based not only on if it can be applied but also if it yields a valid result.

File: spec/models/carto/layer_spec.rb (hunk 13 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: I've changed the naming anyway, I think it's clearer now anyway :)

File: spec/models/carto/layer_spec.rb (hunk 10 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Yeah, make sense! Update incoming!

**Comment by gfiorav** [comment]
Have a second look @juanignaciosl !

Thanks!

File: app/models/carto/layer.rb [hunk 14]
```diff
@@ -241,8 +241,83 @@ def category
       options && options['category']
     end
 
+    def builder?
+      visualization.builder? if visualization
+    end
+
+    def has_valid_source?
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: As both methods are new maybe you can remove `has_valid_source?`.

File: spec/models/carto/layer_spec.rb [hunk 15]
```diff
@@ -82,5 +83,80 @@
         @layer.send(:affected_table_names, sql).should =~ ["cartodb.cdb_tablemetadata", "public.spatial_ref_sys"]
       end
     end
+
+    describe 'source_id recovery' do
+      before(:each) do
+        @layer.options['source'] = @analysis.natural_id
+        @layer.options['previous_source'] = nil
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: I'd replace all external accesses to `@layer.options['previous_source']` with `@layer.previous_source`.

File: app/models/carto/layer.rb [hunk 16]
```diff
@@ -241,8 +241,83 @@ def category
...
+      options['source']
+    end
+
+    def source=(source)
+      self.previous_source = self.source if self.source.present?
+      options['source'] = source
+    end
+
+    def attempt_source_fix
+      strategies = [method(:previous_source),
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: As long as strategies return `nil` or `false` if they don't succeed, this can be replaced with a simpler `source_candidate = previous_source || next_same_letter_source || guessed_source`, isn't it?

File: app/models/carto/layer.rb [hunk 17]
```diff
@@ -241,8 +241,83 @@ def category
...
+    def source
+      options['source']
+    end
+
+    def source=(source)
+      self.previous_source = self.source if self.source.present?
+      options['source'] = source
+    end
+
+    def attempt_source_fix
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Order discussion: I think that you didn't directly answer this in previous discussions. I'm not doubting about this, but I'd like to know the reason why `previous_source` is a better strategy than `next_same_letter_source` and why it is better than `guessed_source`. Forget about performance, as it's not a problem at all. Maybe adding a comment with the algorithm would be nice for our future us :-)

File: app/models/carto/layer.rb [hunk 18]
```diff
@@ -241,8 +241,83 @@ def category
...
+        if valid_source?(source_candidate)
+          self.source = source_candidate
+          break
+        end
+      end
+    end
+
     private
 
+    def next_same_letter_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: This doesn't return the "next" but the "last", isn't it?

File: app/models/carto/layer.rb [hunk 19]
```diff
@@ -241,8 +241,83 @@ def category
...
+      letter = options['letter'] || source.to_s.first
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Does getting `source.to_s.first` make sense here? I recall having commented something related in a previous CR, but it was used at the method body, which doesn't happen here. It's not _bad_, as method name doesn't say anything related to fixing or whatever, so feel free to ignore this if you're sure about it.

File: app/models/carto/layer.rb [hunk 20]
```diff
@@ -241,8 +241,83 @@ def category
...
+
+        letter_analysis_nodes = analysis_nodes.select do |analysis_node|
+          analysis_node.id.first == letter
+        end
+
+        letter_analysis_nodes.map(&:id).sort.last
+      end
+    end
+
+    def guessed_source
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Two things here:
1.- Naming: maybe it should say that this is supposed to return the `previous_source`, not the guessed one. Not sure, though, as there's an exception at 0.
2.- I can't understand the `source` usage. This is supposed to be used as recovery strategy, so `source` is not defined if this method is used at the strategy. So, `number` will always be 0. Maybe this method as an strategy is an overkill?

File: spec/models/carto/layer_spec.rb [hunk 21]
```diff
@@ -82,5 +83,83 @@
...
+
+      it 'looks for highest same letter analysis when no backup' do
+        @layer.options['source'] = bad_source
+        @layer.source.should eq bad_source
+        @layer.options['previous_source'].should be_nil
+
+        @layer.has_valid_source?.should be_false
+        @layer.attempt_source_fix
+
+        @layer.source.should eq @analysis.natural_id
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: :comb: ? :P

File: app/models/carto/layer.rb (hunk 14 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: Sure, I didn't want to share the `valid_source?(source)`method though.

File: app/models/carto/layer.rb (hunk 17 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: `guessed_source` is a last despearte try that will only happen on a layer with no visualization.

I choose to `previous_source` because if a backup exists, it likely used to work. If it doesn't anymore, `next_same_letter_source` is used. No biggie. In the end none will be chosen unless they yield a valid source. If none help, the source won't be changed.

File: app/models/carto/layer.rb (hunk 16 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: No. That is how it was before, remember? :) We decided it was better to test if the result was a valid source and if not go for the next strategy. This could be an array with all the strategies result, but there is no need to precompute all of them if you don't need them yet.

File: app/models/carto/layer.rb (hunk 18 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: Depends on how you look at it! For example, a node "descendants" are those with a smaller number, which makes not intuitive sense since these are "fathers" or "feeders" to those with higher number.

That being said, I do like 'last' more xD. Changing

File: app/models/carto/layer.rb (hunk 19 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: It makes sense as a coalesce to `nil`. `nil.to_s` -> ''

File: app/models/carto/layer.rb (hunk 20 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: 1. Previous source, is already used. It reads a backup in layer#options.
2. Guessed source is guessing that the node missing will have the same letter and the previous number to the faulty source. It's a guess, nothing more. This will only be used on layers without a visualization.

File: spec/models/carto/layer_spec.rb (hunk 15 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: `@layer.previous_source=(source)` forbids you of overwriting a backup with `nil`

File: spec/models/carto/layer_spec.rb (hunk 21 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Haha, sorry! Will add

File: app/models/carto/layer.rb (hunk 16 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Oh, well. I was thinking about checking the validation within the strategy, or adding it within every `||` check as a filter. I really dislike metaprograming for this kind of stuff. Anyway, as I see that you prefer it I'll review it from another point of view ;-) 👍

File: app/models/carto/layer.rb [hunk 22]
```diff
@@ -241,8 +241,83 @@ def category
...
+      options['source'] = source
+    end
+
+    def attempt_source_fix
+      strategies = [method(:previous_source),
+                    method(:last_same_letter_source),
+                    method(:guessed_source)]
+
+      strategies.each do |strategy|
+        source_candidate = strategy.call
```
Comment (gfiorav) [inline app/models/carto/layer.rb]: I wanted to have an array of pointers to methods. I miss C 😢

File: app/models/carto/layer.rb (hunk 20 above)
Comment (gfiorav) [inline app/models/carto/layer.rb]: It could happen that `options['letter']`is empty. If it's there I trust it and if it's not I grab the first char of the faulty source. You could argue its best to always trust the source's first char... not really sure.

File: spec/models/carto/layer_spec.rb (hunk 15 above)
Comment (gfiorav) [inline spec/models/carto/layer_spec.rb]: Ah, sorry, forgot to say: It's private to discourage playing around with the backup from the outside.

File: app/models/carto/layer.rb (hunk 22 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: No, I'm not proposing anything :D. I'm just asking. That's the alternative. I think that we use `send` more often, but your solution is perfectly right.

File: app/models/carto/layer.rb (hunk 20 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Oh, ok. In my mind, "source" was not "faulty" but "missing". My bad. Looks ok 👍

File: spec/models/carto/layer_spec.rb (hunk 15 above)
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: Ok, but as `options['previous_source']` is still available that's not a great security xD. Ok anyway.

**Comment by gfiorav** [comment]
@xavijam can we kill this? 😁

**Comment by xavijam** [comment]
Yes sir! :D
